### PR TITLE
Pagination Cursors on Connections (Relay Specification)

### PR DESCRIPTION
**Pull request by litewarp**
# Description

**UPDATE** This PR only contains the first part of the Relay implementation, the pagination cursors. For the node resolution part, see #285.

Added a bunch of features to make the library fully-compatible with Relay. That is:

- Connections now return totalCount, edges, and pageInfo

- Connections now accept pagination arguments using Relay's cursor-based approach while still allowing for limit/offset calculations.

- A Node interface is added which ensures that all types implementing the interface have an id field that is unique

- A root Query field "node" is added allowing users to query a node by its ID and resolve the correct type based on the labels of the node.

**Input**

```graphql

type Actor implements Node {
  id: ID! @id
  name: String
  movies: [Movie!]! @relationship(type: "ACTED_IN", properties: "ActedIn", direction: OUT)
}

type Movie implements Node {
  id: ID! @id
  title: String
  actors: [Actor!]! @relationship(type: "ACTED_IN", properties: "ActedIn", direction: IN)
}

interface ActedIn {
  screenTime: Int!
}
```

**Output**

```graphql
# only relevant types shown
# ...

type ActorMoviesConnection {
  edges: [ActorMoviesRelationship!]!
  pageInfo: PageInfo!
  totalCount: Int!
}

input ActorMoviesConnectionOptions {
  sort: [ActorMoviesConnectionSort!]
  first: Int
  after: Cursor
}

"""Opaque cursor used for pagination (Relay)"""
scalar Cursor

type MovieActorsConnection {
  edges: [MovieActorsRelationship!]!
  pageInfo: PageInfo!
  totalCount: Int!
}

input MovieActorsConnectionOptions {
  sort: [MovieActorsConnectionSort!]
  first: Int
  after: Cursor
}

type MovieActorsRelationship implements ActedIn {
  cursor: Cursor!
  node: Actor!
  screenTime: Int!
}

"""Pagination information (Relay)"""
type PageInfo {
    hasNextPage: Boolean!
    hasPreviousPage: Boolean!
    startCursor: Cursor
    endCursor: Cursor
}

"""Globally-identifiable node (Relay)"""
interface Node {
  id: ID!
}

type Query {
  actors(where: ActorWhere, options: ActorOptions): [Actor!]!
  movies(where: MovieWhere, options: MovieOptions): [Movie!]!
  node(id: ID!): Node!
}
# ...

```

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [X] TCK tests have been updated
- [X] Integration tests have been updated
- [ ] Example applications have been updated
- [X] New files have copyright header
- [X] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by darrellwarde** [comment]
First comment is, can you please complete the CLA (https://neo4j.com/developer/cla/) so that we can get the pipeline running and see where we stand? Thanks.

File: packages/graphql/src/schema/connection.ts
Comment (darrellwarde) [inline packages/graphql/src/schema/connection.ts:55]: Trying to figure out what's going in here - is `arraySlice` already fully filtered from the database, or is this filtering it further?

**Review by darrellwarde — COMMENTED** [review]
This seems like really great work, thanks for the submission. Honestly, this looks like it should work in practice, but there are a couple of things I would like to see clarified/fixed:

- I believe all node types should implement the `Node` interface, but this introduces a lot of problems with enforcing the `id` field, and migrating existing datasets. This was always one of our blockers for doing this now.
- How are cursors calculated? Trying to figure this out as it's not overly clear, and I want to make sure there are no downsides to this approach. We were considering actually storing cursors on relationships as properties.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:123]: I had thought there might be some benefit to abstracting it in case there was a transform to be done, but after diving even deeper into the [spec](https://relay.dev/assets/files/connections-932f4f2cdffd79724ac76373deb30dc8.htm#sec-Edge-Types.Introspection), it says that using String is fine.

I'll change it in the next commit.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:184]: So this is a little tricky. It actually _does _ use `composer.createInterfaceTC`, but further down around line 1142 where you iterate through the `interfaces` array.

In order to pass the `checkNodeImplementsInterfaces` function without throwing, the InterfaceTypeDefinitionNode for the Node interface needs to be added to the `interfaces` array. So I created the definition node, added it to the interfaces array and then it gets added to the schema through `composer.createInterfaceTC`.

I agree with you that ideally we could just define the interface first and then output the definition, but I was not sure how to use the resulting graphql-compose [InterfaceTypeComposer](https://graphql-compose.github.io/docs/api/InterfaceTypeComposer.html) class to generate the appropriate InterfaceTypeDefinitionNode to pass to the `checkNodeImplementsInterface` function.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:175]: Ok that makes sense. I can see how it would be tricky to migrate an existing data set. 

One thing to note is that the Relay specification has a few steadfast rules, and having an id property on all implementing nodes [is required by the specification](https://graphql.org/learn/global-object-identification/#introspection). 

One thought I had was to make the changes by this PR (i.e. full relay implementation) optional that you could set with a flag in the config. I believe I recall seeing that in one of the Relationship-Properties proposals.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:1042]: Good call. Will change in the next commit.

File: packages/graphql/src/schema/connection.ts
Comment (litewarp) [reply packages/graphql/src/schema/connection.ts:26]: So the `createConnectionWithEdgeProperties` function *is* an implementation of that algorithm. 

I'm not knowledgeable enough about cypher to know how to translate the algorithm for direct use. That said, with the exception of the `offsetToCursor` and `cursorToOffset` functions (see more discussion below), everything is pure math and logic checks. So it *should* be doable. Alas, despite lurking in the "#cypher" channel, I would need help with this

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:1155]: Yes, I would need some help here. 🙂

The trick for me is that I don't know how to build the full projected query *without* knowing the type name. 

Any ideas?

File: packages/graphql/src/schema/connection.ts
Comment (litewarp) [reply packages/graphql/src/schema/connection.ts:55]: I think this line is actually unnecessary.  This function is a replication of grapqhl-relay-js [connectionFromArraySlice](https://github.com/graphql/graphql-relay-js/blob/fa978fbfaa367b79b4616d091471be87e84f3dbe/src/connection/arrayConnection.js#L92) with the following change:

```Typescript
const edges = slice.map((value, index) => ({
  ...value, # <= changed from node: value,
  cursor: offsetToCursor(startOffset + index)
})
``` 

As a result, I was reluctant to start changing it, but you are correct that the cypher statement already returns a data slice based on limit/offset, so it's a redundant check.

**Comment by litewarp** [comment]
> * How are cursors calculated? Trying to figure this out as it's not overly clear, and I want to make sure there are no downsides to this approach. We were considering actually storing cursors on relationships as properties.

So the cursors are created based on a [base64 transform](https://github.com/graphql/graphql-relay-js/blob/main/src/utils/base64.js) exported as the helpers below by graphql-relay-js(https://github.com/graphql/graphql-relay-js/blob/fa978fbfaa367b79b4616d091471be87e84f3dbe/src/connection/arrayConnection.js#L132):  

```Typescript
const PREFIX = 'arrayconnection:';

/**
 * Creates the cursor string from an offset.
 */
export function offsetToCursor(offset: number): ConnectionCursor {
  return base64(PREFIX + offset.toString());
}

/**
 * Extracts the offset from the cursor string.
 */
export function cursorToOffset(cursor: ConnectionCursor): number {
  return parseInt(unbase64(cursor).substring(PREFIX.length), 10);
}

```

However, this is only one method of generating cursors, although most of the implementations I've seen seem to use base64:  
- [graphql-compose-connection](https://github.com/graphql-compose/graphql-compose-connection/blob/master/src/cursor.ts). 
- [graphql-relay-go](https://github.com/authgear/graphql-go-relay/blob/df672205b892b3a16476b3738c6f33e54d49237e/array_connection.go#L126) 
- [nexus relay connection plugin](https://github.com/graphql-nexus/nexus/blob/8e0ce1b4a156cd50b061b47c8a1241729b4dd2a3/src/plugins/connectionPlugin.ts#L349)

The key thing is that cursors should be translatable in some way to the offset so that the database call doesn't have to traverse the entire collection to find the subsequent starting point when the user calls for the next group of data.

File: packages/graphql/src/schema/connection.ts
Comment (litewarp) [reply packages/graphql/src/schema/connection.ts:26]: So it looks like apoc already has base64 [encode](https://neo4j.com/labs/apoc/4.2/overview/apoc.text/apoc.text.base64Encode/) and [decode](https://neo4j.com/labs/apoc/4.2/overview/apoc.text/apoc.text.base64Decode/) procedures. 

So yes, I think it will be possible to write a cypher algorithm that does this. But this is probably something saved for one of your expertise.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:169]: So I think the solution to this is to use id encoding. All we would need from users is an identification of a unique field property. Then you can encode the TypeName and the uniqueValue into a string that you use for the id. This way the id is globally unique amongst types, you can auto-generate it based on identified unique keys, etc.

For example, using the [graphql-relay-js](https://github.com/graphql/graphql-relay-js/blob/fa978fbfaa367b79b4616d091471be87e84f3dbe/src/node/node.js#L87) library again as example, there are two helpers: 

```Typescript
/**
 * Takes a type name and an ID specific to that type name, and returns a
 * "global ID" that is unique among all types.
 */
export function toGlobalId(type: string, id: string | number): string {
  return base64([type, GraphQLID.serialize(id)].join(':'));
}

/**
 * Takes the "global ID" created by toGlobalID, and returns the type name and ID
 * used to create it.
 */
export function fromGlobalId(globalId: string): ResolvedGlobalId {
  const unbasedGlobalId = unbase64(globalId);
  const delimiterPos = unbasedGlobalId.indexOf(':');
  return {
    type: unbasedGlobalId.substring(0, delimiterPos),
    id: unbasedGlobalId.substring(delimiterPos + 1),
  };
}
```

So long as the id value is unique across the Node, by encoding it with the typeName, it will be unique across all types.

For users that don't have an id field, or want to use another field, you can autogenerate it for them. All you'll need them to do is identify the field with the @id directive?

```graphql
# using uuid
type Movie {
  id: ID!

  _id: ID @id
}

# using other field
type Actor {
  id: ID!
 
  imdbNo: String! @id(autogenerate: false) 
}
``` 

This would also make node resolution easier since all you'll have to do is decode the id, so two database trips aren't required.

File: packages/graphql/src/schema/connection.ts
Comment (litewarp) [reply packages/graphql/src/schema/connection.ts:26]: This was my best attempt at trying to write a cypher statement to create the cursors. I still haven't gotten it to work so no guarantee at all that it's correct. But maybe you or another neo4j all-star could use it as a starting point.

```
Match (this:MOVIE { title: "Point Break" })<-[this_acted_in:ACTED_IN]-(this_actor:Actor)
WITH this, this_acted_in, this_actor
WITH this, collect({ screenTime: this_acted_in.screenTime }) as edges, count(this_acted_in) as totalCount

CALL apoc.text.base64Decode("YXJyYXlDb25uZWN0aW9uNA==") YIELD prefixedCursor
CALL apoc.text.replace(prefixedCursor, "arrayConnection", "") YIELD stringValue
CALL apoc.number.parseInt(stringValue) YIELD intValue

CALL apoc.when(apoc.meta.isType(intValue, "INTEGER"), "RETURN intValue as offset", "RETURN 0 as offset") YIELD value

UNWIND edges as edge
UNWIND totalCount as i
CALL apoc.text.join(["arrayConnection", apoc.convert.toString(i + value.offset)]) YIELD cursorString
CALL apoc.text.base64Encode(cursorString) YIELD cursor
WITH collect(edge{.*, cursor: cursor }) as edges
RETURN { edges: edges, totalCount: size(edges) } AS actorsConnection
```

**Comment by darrellwarde** [comment]
> The key thing is that cursors should be translatable in some way to the offset so that the database call doesn't have to traverse the entire collection to find the subsequent starting point when the user calls for the next group of data.

Thanks for the explanation and references. This makes sense, however one thing I can't quite work out is how this works when data changes between queries. Say you had a query for the first 5 comments of all posts, ordered by their creation timestamp descending:

```graphql
query {
  posts {
    commentsConnection(options: { first: 5, sort: { node: { createdAt: DESC } } }) {
      edges {
        cursor
        node {
          ...commentFields
        }
      }
    }
  }
}
```

Say this returns the following 5 posts in descending order the first time around:

1. Post 5 (cursor: "a")
2. Post 4 (cursor: "b")
3. Post 3 (cursor: "c")
4. Post 2 (cursor: "d")
5. Post 1 (cursor: "e")

Now let's say a 6th post is created, and it is at the top of the list because it is the most recently created, I imagine your list would now look like:

1. Post 6 (cursor: "a")
2. Post 5 (cursor: "b")
3. Post 4 (cursor: "c")
4. Post 3 (cursor: "d")
5. Post 2 (cursor: "e")

I might be completely misunderstanding - I only come to this conclusion because if cursors are calculated based on offset, rather than being "saved" on particular relationships (which is what we thought would have to be done), this is how I imagine it will work. If that is the case, would this not be a problem in that the starting point can change in between queries?

**Comment by darrellwarde** [comment]
> First comment is, can you please complete the CLA (https://neo4j.com/developer/cla/) so that we can get the pipeline running and see where we stand? Thanks.

Great, thanks for doing that! Looks like there is just one test failure because of some more recent changes on the 2.0.0 branch - can you please merge those changes from the 2.0.0 into your branch?

**Comment by darrellwarde** [comment]
Right, I have a bit of a painful request for you, which I hope you won't mind doing.

Would you mind pulling out anything to do with the `Node` interface and the `node` query into a different branch and raise a separate PR for that? I think we can all agree that is where the complexity lies here. If we can get that set up, I think we will be able to get the cursor-based pagination side of things reviewed and merged pretty quickly, and we can better focus on how to add support for the node IDs, etc. 🙂

File: packages/graphql/src/schema/connection.ts
Comment (darrellwarde) [reply packages/graphql/src/schema/connection.ts:26]: Honestly, I probably wouldn't have done much better! 😝  There's definitely a lot of complexity there, and it would increase our reliance on APOC, which we're trying to avoid. Let's stick with the function implementation for the time-being, and keep this in our back pocket!

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (darrellwarde) [reply packages/graphql/src/schema/make-augmented-schema.ts:1149]: We can continue this discussion if and when the "node" side of things gets moved into a different branch and PR, but indeed, the projection is tricky! You may actually be right that it needs to be done in two queries... 🙁

**Comment by litewarp** [comment]
> > The key thing is that cursors should be translatable in some way to the offset so that the database call doesn't have to traverse the entire collection to find the subsequent starting point when the user calls for the next group of data.
> 
> Thanks for the explanation and references. This makes sense, however one thing I can't quite work out is how this works when data changes between queries. Say you had a query for the first 5 comments of all posts, ordered by their creation timestamp descending:
> 
> ```graphql
> query {
>   posts {
>     commentsConnection(options: { first: 5, sort: { node: { createdAt: DESC } } }) {
>       edges {
>         cursor
>         node {
>           ...commentFields
>         }
>       }
>     }
>   }
> }
> ```
> 
> Say this returns the following 5 posts in descending order the first time around:
> 
> 1. Post 5 (cursor: "a")
> 2. Post 4 (cursor: "b")
> 3. Post 3 (cursor: "c")
> 4. Post 2 (cursor: "d")
> 5. Post 1 (cursor: "e")
> 
> Now let's say a 6th post is created, and it is at the top of the list because it is the most recently created, I imagine your list would now look like:
> 
> 1. Post 6 (cursor: "a")
> 2. Post 5 (cursor: "b")
> 3. Post 4 (cursor: "c")
> 4. Post 3 (cursor: "d")
> 5. Post 2 (cursor: "e")
> 
> I might be completely misunderstanding - I only come to this conclusion because if cursors are calculated based on offset, rather than being "saved" on particular relationships (which is what we thought would have to be done), this is how I imagine it will work. If that is the case, would this not be a problem in that the starting point can change in between queries?

So it depends on how your querying is done. Every time there's a change in the query (i.e., sort, or new filters, or new where), cursors will be recalculated. So if there's a new node, when the client refreshes the data, yes, the list will be repaginated.

In terms of "the starting point changing", this would only be an issue on the client when they have not refetched new data (causing recalculation of the cursors). In that case, if you don't want to cause a refetch after a mutation, you can actually manually add the new item to the cache while keeping the cursors consistent by using the Relay [connection handler](https://relay.dev/docs/guided-tour/list-data/updating-connections/#adding-edges). 

I suppose I don't really understand how storing "cursors" on the relationship lets you do pagination. If a user resorts, you'd have to recreate the entire list of cursors in the order they wanted. And then when you want to start off after a certain cursor, you'd need to traverse through the entire list to find the starting point.

File: packages/graphql/src/schema/connection.ts
Comment (darrellwarde) [inline packages/graphql/src/schema/connection.ts:52]: It's obviously going to do no harm, but it seems there's no real reason to keep any code related to `last` or `before` here, given that we're going with `first` and `after`?

**Review by darrellwarde — CHANGES_REQUESTED** [review]
I feel like we're close on this one now that this PR just focuses on pagination. A few points for review, would especially like to see some integration testing for the pagination functionality. Thanks again for the work on this, really great to see! 🙂

**Comment by darrellwarde** [comment]
> > > The key thing is that cursors should be translatable in some way to the offset so that the database call doesn't have to traverse the entire collection to find the subsequent starting point when the user calls for the next group of data.
> > 
> > 
> > Thanks for the explanation and references. This makes sense, however one thing I can't quite work out is how this works when data changes between queries. Say you had a query for the first 5 comments of all posts, ordered by their creation timestamp descending:
> > ```graphql
> > query {
> >   posts {
> >     commentsConnection(options: { first: 5, sort: { node: { createdAt: DESC } } }) {
> >       edges {
> >         cursor
> >         node {
> >           ...commentFields
> >         }
> >       }
> >     }
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Say this returns the following 5 posts in descending order the first time around:
> > 
> > 1. Post 5 (cursor: "a")
> > 2. Post 4 (cursor: "b")
> > 3. Post 3 (cursor: "c")
> > 4. Post 2 (cursor: "d")
> > 5. Post 1 (cursor: "e")
> > 
> > Now let's say a 6th post is created, and it is at the top of the list because it is the most recently created, I imagine your list would now look like:
> > 
> > 1. Post 6 (cursor: "a")
> > 2. Post 5 (cursor: "b")
> > 3. Post 4 (cursor: "c")
> > 4. Post 3 (cursor: "d")
> > 5. Post 2 (cursor: "e")
> > 
> > I might be completely misunderstanding - I only come to this conclusion because if cursors are calculated based on offset, rather than being "saved" on particular relationships (which is what we thought would have to be done), this is how I imagine it will work. If that is the case, would this not be a problem in that the starting point can change in between queries?
> 
> So it depends on how your querying is done. Every time there's a change in the query (i.e., sort, or new filters, or new args), cursors will be recalculated on the server and the client will update with those new cursors. In the world where there are no changes, relay knows the order of the list and the number of items so recalculating the position on the client side is doable.
> 
> So in terms of "the starting point changing", this would only be an issue on the client when they have not refetched new data (causing recalculation of the cursors). In that case, if you don't want to cause a refetch after a mutation, you can actually manually add the new item to the cache while keeping the cursors consistent by using the Relay [connection handler](https://relay.dev/docs/guided-tour/list-data/updating-connections/#adding-edges).
> 
> If you're using Apollo, they have a field policy [helper](https://www.apollographql.com/docs/react/pagination/cursor-based/#relay-style-cursor-pagination) specifically for relay pagination. I believe urql has similar functionality.
> 
> I suppose I don't really understand how storing "cursors" on the relationship lets you do pagination. If a user resorts, you'd have to recreate the entire list of cursors in the order they wanted. And then when you want to start off after a certain cursor, you'd need to traverse through the entire list to find the starting point.

I guess I'm less thinking about examples where a user is querying the data that they're mutating themselves, but where data has been mutated by another user (another client, so cache wouldn't have updated unless subscriptions are being used, etc.). In this case, the offset could be "changed" between queries by another user inserting data, so when our user resumes pagination from a known cursor, with the same filters, the node that this cursor relates to has now been shifted to the right.

I might be misunderstanding and it's very hard to express my thoughts through text - you seem to have significantly more knowledge of Relay than I do! 🤯 

The thinking of having cursors as properties on relationships was in very early stages, so it might have been completely off the mark. But it was less thinking that cursors were related to offsets, so that if more data was inserted at the start of the list, you still get the same node back with the cursor, compared to a particular position in a list with this approach. Yes, you would have to fetch the entire list matching the filter and then slice it down based on the `first` and `after` arguments, but this could all easily be done in Cypher which means the Cypher planner should have been able to deal with it quite well.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:978]: Good question. So I think the arguments need to be moved up to be direct arguments of the field. There is nothing in the specification, but if you check out the relay-compiler (used with the Relay client), it'll [throw an error](https://github.com/facebook/relay/blob/main/packages/relay-compiler/transforms/ConnectionTransform.js#L698) if it can't find arguments directly on the field.

I also could not find any examples where the arguments were nested. A person once raised the [issue](https://github.com/facebook/relay/issues/3285) but there was no comment or resolution from FB and the person closed it without any advice.

I'll move them out of the options object on next go. Any thoughts if I should hoist the sort argument as well so everything is directly on the field?

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (litewarp) [reply packages/graphql/src/schema/make-augmented-schema.ts:127]: Got it. Will update in the next commit.

File: packages/graphql/tests/integration/connection-resolvers-int.test.ts
Comment (litewarp) [reply packages/graphql/tests/integration/connection-resolvers-int.test.ts:26]: Makes sense. Will add to it.

File: packages/graphql/src/schema/connection.ts
Comment (litewarp) [reply packages/graphql/src/schema/connection.ts:52]: Nope. We should rip it out. It's only likely to cause confusion. Will also be in the next commits.

File: packages/graphql/src/schema/make-augmented-schema.ts
Comment (darrellwarde) [reply packages/graphql/src/schema/make-augmented-schema.ts:978]: Sure, so definitely hoist up `first` and `after`.

`sort` is tricky! If we hoist it up, it becomes inconsistent with what is on a "normal" relationship field, but you could argue that it's already inconsistent what with `first` and `after` at the root. But then if we leave it, it's sat there in `options` all by itself which doesn't make an awful lot of sense!

My gut says also hoist it up and bin off the `ConnectionOptions` types, as `where` and `sort` go quite well together. However, would like to get some thoughts off @danstarns and/or @oskarhane before we make a decision here.

Comment (darrellwarde) [reply packages/graphql/src/schema/make-augmented-schema.ts:978]: Had a chat with @danstarns, I think go ahead and also hoist up `sort` as well, as we're already inconsistent with where the pagination arguments sit. Also means we can get rid of the `ConnectionOptions` type which is nice. 

You're also going to want to merge from the upstream 2.0.0 branch as we've merged a couple of branches in and it looks like the tests are a bit broken again.

**Comment by litewarp** [comment]
Just an FYI. There was a bug with the skipLimitStr function. The correct syntax (as I learned on the discord today) for pagination is [offset..sliceEnd] rather than [offset..limit]. Also since the offsets start at 0 but the cypher string should start with 1, you need to make sure the create-connection-and-params translation increments it by one. 

I ended up writing some unit tests for the pagination utilities to help me debug and ensure consistency in the output.

File: packages/graphql/tests/integration/pagination.int.test.ts
Comment (darrellwarde) [inline packages/graphql/tests/integration/pagination.int.test.ts:1]: These strike me more as unit tests which should be living at the same level as the source code, i.e. `pagination.test.ts` in the same directory as `pagination.ts`.

**Review by darrellwarde — CHANGES_REQUESTED** [review]
This is really starting to look good now, the only comment that I really have is regarding moving and renaming a test file! I've also asked @danstarns to review this, but it feels to me like we're pretty close here. 🙂

File: packages/graphql/src/schema/pagination.ts
Comment (darrellwarde) [inline packages/graphql/src/schema/pagination.ts:86]: This should probably be in the `translate` directory, but we can easily shuffle things around when merged, so not a blocker.

**Review by darrellwarde — APPROVED** [review]
Only one small comment from me now, but this is easily fixed at any point, and everything else looks good to me!

Thanks for your work on this @litewarp, saved us a lot of time and implemented a much requested feature. Will wait for a review from @danstarns before merging.

**Comment by danstarns** [comment]
Hey @litewarp, Am I doing something wrong? 

Schema: 

```graphql
type Person {
    name: String
    actedIn: [Movie] @relationship(type: "ACTED_IN", direction: OUT)
}

type Movie {
    title: String
}
```

Query: 

```graphql
 {
  people {
    name
    actedInConnection {
      edges {
        cursor
        node {
          title
        }
      }
    }
  }
}
```

Error: 
```
{
  "errors": [
    {
      "message": "Cannot read property 'cursor' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "people",
        4,
        "actedInConnection"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'cursor' of undefined",
            "    at Object.createConnectionWithEdgeProperties (C:\\code\\neo4j\\graphql\\packages\\graphql\\src\\schema\\pagination.ts:57:36)",
            "    at resolve (C:\\code\\neo4j\\graphql\\packages\\graphql\\src\\schema\\make-augmented-schema.ts:997:32)",
            "    at field.resolve (C:\\code\\neo4j\\graphql\\node_modules\\apollo-server-core\\src\\utils\\schemaInstrumentation.ts:103:18)",
            "    at resolveField (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:464:18)",
            "    at executeFields (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:292:18)",
            "    at collectAndExecuteSubfields (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:748:10)",
            "    at completeObjectValue (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:738:10)",
            "    at completeValue (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:590:12)",
            "    at completeValue (C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:556:21)",
            "    at C:\\code\\neo4j\\graphql\\node_modules\\graphql\\execution\\execute.js:620:25"
          ]
        }
      }
    }
  ],
  "data": null
}
```

**Comment by litewarp** [comment]
You need to add properties to the relationship. I was actually going to raise this in another PR but the library doesn't create a connection unless there are properties on the relationship.

**Comment by danstarns** [comment]
> You need to add properties to the relationship. I was actually going to raise this in another PR but the library doesn't create a connection unless there are properties on the relationship.

I'm not sure about this, seeming that most of our users are currently not using relationship properties this would mean they couldn't use the cursor. Is this something that can be resolved ? and if so what do you propose?

**Comment by darrellwarde** [comment]
> You need to add properties to the relationship. I was actually going to raise this in another PR but the library doesn't create a connection unless there are properties on the relationship.

I think this was done to minimise the size of the schema, because before this PR there was literally zero value in having the connection types unless relationship properties were defined, but that's obviously all changed now! So we can either do that here, or do it afterwards - my gut says afterwards so that we can get this merged.

**Review by danstarns — APPROVED** [review]
Nice contribution 🍺

**Pull request merged**
